src/gateways/discord_gateway.ts

```diff
@@ -0,0 +1,57 @@
+import { Client, GuildMember, Guild, User } from "discord.js";
+
+
+export class DiscordGateway {
+    private static instance: DiscordGateway;
+
+    private client: Client;
+
+    private constructor() {
+        this.client = new Client();
+        this.client.login(process.env.BOT_TOKEN);
+
+        this.client.on("error", (error: Error) => {
+            console.error("Catastrophic error occurred on Discord.js client", error);
+            console.error("Exiting");
+            process.exit(1);
+        });
+
+        this.client.on("disconnect", () => {
+            console.error("server disconnected...");
+        });
+    }
+
+    static get getDiscordClient() {
+        if (!DiscordGateway.instance) {
+            DiscordGateway.instance = new DiscordGateway();
+        }
+
+        return DiscordGateway.instance.client;
+    }
+
+    static get getInstance() {
+        if (!DiscordGateway.instance) {
+            DiscordGateway.instance = new DiscordGateway();
+        }
+
+        return DiscordGateway.instance;
+    }
+
+    setPresenceUpdateHandler(handler: (oldMember: GuildMember, newMember: GuildMember) => Promise<void>) {
+        this.client.on("presenceUpdate", handler);
+    }
+
+    getAllAvailableGuilds() {
+        console.info("Getting all available guilds");
+        return DiscordGateway.instance.client.guilds.map((value: Guild) => value);
+    }
+
+    getViableVoiceChannelForUser(user: User): GuildMember {
+        const guilds: Guild[] = DiscordGateway.instance.getAllAvailableGuilds().filter(guild => !!guild.member(user).voiceChannel);
+        if (guilds.length) {
+            return guilds[0].member(user);
+        } else {
+            return undefined;
+        }
```

Comment:
this can be simplified to
```javascript
const [guild]: Guild = DiscordGateway.instance.getAllAvailableGuilds()
    .filter(guild => !!guild.member(user).voiceChannel);
return guild;
```

Reply:
since if array length is zero, it will return undefined, which is the expected behavior.

Reply:
actually, we need to have GuildMember, not Guild...